app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.h

```diff
@@ -1,160 +1,191 @@
-        Internal, //!< Refers to the the internal display
-        Null, //!< Refers to the null display which is used for discarding data
-    };
-
-    enum class LayerStatus {
-        Uninitialized, //!< The layer hasn't been initialized
-        Stray, //!< The layer has been initialized as a stray layer
-        Managed, //!< The layer has been initialized as a managed layer
+    struct BufferSlot {
+        BufferState state{BufferState::Free};
+        u64 frameNumber{}; //!< The amount of frames that have been queued using this slot
+        bool wasBufferRequested{}; //!< If GraphicBufferProducer::RequestBuffer has been called with this buffer
+        bool isPreallocated{}; //!< If this slot's graphic buffer has been preallocated or attached
+        std::shared_ptr<gpu::Texture> texture{};
+        std::unique_ptr<GraphicBuffer> graphicBuffer{};
     };
 
     /**
-     * @brief IGraphicBufferProducer is responsible for presenting buffers to the display as well as compositing and frame pacing
-     * @url https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp
+     * @brief An endpoint for the GraphicBufferProducer interface, it approximately implements BufferQueueProducer but also implements the functionality of interfaces called into by it such as GraphicBufferConsumer, Gralloc and so on
+     * @url https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/include/gui/IGraphicBufferProducer.h
+     * @url https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/libs/gui/IGraphicBufferProducer.cpp
+     * @url https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/include/gui/BufferQueueProducer.h
+     * @url https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/libs/gui/BufferQueueProducer.cpp
+     * @url https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/include/gui/BufferQueueCore.h
+     * @url https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/libs/gui/BufferQueueCore.cpp
      */
     class GraphicBufferProducer {
       private:
         const DeviceState &state;
-        std::unordered_map<u32, std::shared_ptr<Buffer>> queue; //!< A vector of shared pointers to all the queued buffers
+        std::mutex mutex; //!< Synchronizes access to the buffer queue
+        constexpr static u8 MaxSlotCount{16}; //!< The maximum amount of buffer slots that a buffer queue can hold, Android supports 64 but they go unused for applications like games so we've lowered this to 16 (https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/include/gui/BufferQueueDefs.h;l=29)
+        std::array<BufferSlot, MaxSlotCount> queue;
+        u8 activeSlotCount{}; //!< The amount of slots in the queue that can be dequeued
```

Comment:
use an i32 for these things, android does so it makes sense to

Reply:
I use `unsigned` as a guarantee here, the code will break very hard if any of these values are negative

Reply:
well im sure android's code would too? if not use u32

Reply:
There's no need to make the type wider either since `MaxSlotCount` is `u8` and this is guaranteed to never go above that, all this would do is make the loads/stores go from `LDRB`/`STRB` to `LDR`/`STR` which are marginally less efficient but for no reason, none of the arithmetic on the CPU side uses this directly so no instructions will be generated from the usage of it here.

Reply:
but android itself supports more slots than a u8 no?

Reply:
It actually doesn't, Android just likes to be more inefficient I suppose?

Reply:
Word

Reply:
> none of the arithmetic on the CPU side uses this directly so no instructions will be generated from the usage of it here.
Anyway, this is sort of pedantic so we'll address it if it comes up in the profiler till then we can take advantage of the lower memory footprint :P